### ISSUE TITLE
Shuffle dungeon rewards

### ISSUE THREAD

**Issue by fenhl** [TestRunnerSRL/OoT-Randomizer#1833]
This is a duplicate of #229 and #544, which were closed in favor of Triforce Hunt due to the complexities outlined below. The community has learned more about the game since then and the feature seems within the realm of the possible now, so I am opening this issue as an overview of what still needs to be done for it to land. Anyone opposed to the setting being added may want to read [this defence by Revven](https://discord.com/channels/274180765816848384/438697911753113600/975581428052029450).

# The feature

A new setting is added (probably at the top of the “Shuffle Dungeon Items” section) that controls where dungeon rewards (Medallions and Spiritual Stones) can be found. The current behavior, i.e. shuffled among each other, becomes the option “Dungeon Reward Locations”, and the other options are the usual ones for dungeon items: “Vanilla Locations”, “Own Dungeon”, “Regional”, “Overworld Only”, “Any Dungeon”, and “Anywhere”. If one of the new options (other than “Vanilla Locations”) is enabled, each blue warp grants a randomized item the first time it is used.

## `Links Pocket`

The dungeon reward location corresponding to the vanilla Light Medallion, currently called `Links Pocket`, is special in that it is currently always a “skipped” location, similar to the locations skipped by the “Skip Child Zelda” and “Pre-completed Dungeons” settings. If this behavior were to be kept, shuffling dungeon rewards would start each seed with a random item (which may be junk), so I think this is also a good opportunity to consider changing the behavior. Some possibilities:

* Remove the special handling and return the check to the `Beyond Door of Time` region. For convenience, instead of requiring a Master Sword pull (which means two long cutscenes if you're already adult), we may want to change the check to simply be given to Link if he's in the region, similar to how `Sheik at Temple` works. Renaming the location to something like `Reward from Rauru` would make sense.
* Make this change only when dungeon rewards are shuffled.
* Add a setting controlling whether the location is skipped, which is disabled when dungeon rewards aren't shuffled. I'm told this is what OoT3DR does.
* Just leave it up to a setting regardless of dungeon reward shuffle. This is my preferred option since it allows the most flexibility, people have expressed interest in a “one random starting item” mode recently, and there's no technical reason I'm aware of for gating the change in behavior behind dungeon reward shuffle.

# Components

- [x] The fill algorithm needs to be adjusted to handle dungeon rewards and their locations as regular items, while still shuffling them before entrances if they're on vanilla/reward locations. This can be achieved by treating them like regular dungeon items and skipping the `fill_bosses` call in `Main`.
- [x] Medallion and Spritual Stone models need to be added as regular item models for freestandings and the get item cutscene. Working thanks to the splitter script from OoTMM and a PR from @mracsys.
- [x] Blue warps need to be modified to give normal items when dungeon rewards are shuffled. See also #265.
- [x] Dungeon rewards need to be defined as major items in the randomizer code. See also #1730.
- [ ] The `Sheik in Kakariko` cutscene currently only triggers when you enter Kakariko Village, not when you're already there (e.g. when receiving the last of the 3 required medallions from Anju or from another world). We need to decide if we want to change this behavior. We might also want to test `Sheik at Temple` to make sure it works correctly. (LACS and the rainbow bridge are presumably fine since there are already settings where you can receive the last required item while inside the trigger. The OoT cutscene in Hyrule Field has been tested and works.) Tracked as #2226.
- [ ] We need to decide what do do with the `Links Pocket` location, as outlined above, and implement the decision. Tracked as #2225.
- [x] We should consider adding dual hints for boss hearts and blue warps.
- [x] Goals need to be adjusted when dungeon rewards are shuffled.
- [x] Plentiful and ludicrous item pools should add extra dungeon rewards as appropriate.
- [x] The item model in the Big Octo cutscene needs to be updated to support non-dungeon-reward items.
- [x] We may want to treat compasses that give info like misc. item hint locations, counting dungeon rewards as “already hinted” if the compass is reachable without the reward.

I have a work-in-progress version of dungeon reward shuffle on [my branch](https://github.com/fenhl/OoT-Randomizer). Components checked off above are already implemented there.

**Comment by fenhl** [comment]
I translated the file splitting script from OoTMM into Python at https://github.com/fenhl/OoT-Randomizer/tree/dev-fenhl-medallion-models but haven't verified whether it works correctly. **Flashing lights warning** for freestanding dungeon rewards, presumably due to the missing draw table entry.

**Comment by RawZ06** [comment]
I have a question about this feature, it can be possible to preserve actual bridge goal by splitting it with medaillons/stones and beat bosses ? The random reward impact only songs : Song from Ocarina of Time, Sheik in Kakariko and Sheik at Temple of Time and LACS but we need to beat some bosses to open bridge to Ganon Tower.

**Comment by fenhl** [comment]
Adding a bridge condition for defeating bosses should be possible if the dev team wants to include that as a feature.

**Comment by NeunEinser** [comment]
Isn't this it already?
![grafik](https://user-images.githubusercontent.com/12124394/209414748-8f4837d1-2f03-4935-b0f5-1216df5a16a7.png)

**Comment by fenhl** [comment]
This is collecting dungeon rewards, not defeating bosses. The option name should probably be clarified when dungeon reward shuffle is added.

**Comment by cjohnson57** [comment]
In regard to Link's Pocket, here's my opinion:
* Change it to be received by being in the beyond DoT region, as others have said it'd be annoying to swap ages just for the check
* This part I'm less certain about, but I think I'm in agreeance with making it its own dropdown setting where you can choose to either:
    * Move the check to beyond DoT (default)
    * Keep the check in Link's pocket 
    * Keep it in his pocket, and guarantee a dungeon reward
 
Even with reward shuffle, I imagine there are many who'd like to still get a free starting reward. Ofc, with reward shuffle off, the latter 2 options are equivalent 

And my reasoning for the default is that the whole pocket thing was basically invented as a convenience of the randomizer and difficulty of adding the check, to a new player they'd probably see no reason to keep the pocket check as the beyond DoT check is closer to what the check "should" be

As for the dual hints, I think that would be good for sure, since it's basically the same check for 2 items.

**Comment by fenhl** [comment]
A quick change I made on my branch just now: Dungeon rewards are added to the list of unhintable woth/goal items alongside Triforce pieces and gold skulltula tokens. This only matters if the dungeon rewards aren't already hinted (i.e. maps/compasses give info and are removed, or maps/compasses _don't_ give info and altar misc. hint is off). I figure someone playing without the compass/altar hints probably also doesn't want the rewards to be hinted directly.

**Comment by r0bd0g** [comment]
I want to mention, last I checked "The Sheik in Kakariko cutscene currently only triggers when you enter Kakariko Village" doesn't actually work if you enter from a grotto, last I checked? but admittedly it was a while ago. (I guess technically it could be a problem if all of your exits from Kak are one way?)

**Comment by fenhl** [comment]
For anyone curious about the status of this feature, the current plan is to open a pull request once #1820 is merged.

**Comment by fenhl** [comment]
> I want to mention, last I checked "The Sheik in Kakariko cutscene currently only triggers when you enter Kakariko Village" doesn't actually work if you enter from a grotto, last I checked? but admittedly it was a while ago. (I guess technically it could be a problem if all of your exits from Kak are one way?)

This is still the case.
